### ISSUE TITLE
Ignore trailing semicolon after closing brace of custom property set block in no-extra-semicolons

### ISSUE THREAD

**Issue by kevinSuttle** [stylelint/stylelint#1830]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

bug where `no-extra-semicolons` and `block-closing-brace-newline-after` are triggered on an `@apply` block nested in `:root`.

> Which rule, if any, is this issue related to?

`no-extra-semicolons`, `block-closing-brace-newline-after`

> What CSS is needed to reproduce this issue?

```css

/* stylelint-disable root-no-standard-properties, selector-type-case, no-extra-semicolons */
:root {
  --ThemeButtonStyles: {
    @apply --ThemeAnimationStyles;

    border-width: 1px;
    font-family: var(--LightFontFamily);
    font-size: .875em;
    cursor: pointer;
    pointer-events: all;
    height: var(--UIComponentSize);
    box-sizing: border-box;
    margin: 0;
    transition-property: background-color;
  }; /* triggered here: no-extra-semicolons, block-closing-brace-newline-after */

  --ThemeButtonLabelProperties: {
    font-family: var(--LightFontFamily);
    color: currentColor;
  }; /* triggered here: no-extra-semicolons */
}
```

> What stylelint configuration is needed to reproduce this issue?

```json
{
  "extends": "stylelint-config-standard",
  "rules": {
    "custom-property-no-outside-root": true,
    "declaration-block-no-duplicate-properties": true,
    "no-eol-whitespace": true,
    "number-leading-zero": "never",
    "root-no-standard-properties": true,
    "selector-root-no-composition": true,
    "selector-type-no-unknown": [true, {
    	"ignoreTypes": ["/^--/", "--Theme"]
	   }]
  }
}
```

> Which version of stylelint are you using?

`7.1.0`

> How are you running stylelint: CLI, PostCSS plugin, Node API?

Sublime plugin

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

Not that I know of.

> What did you expect to happen?

"No warnings to be flagged."

> What actually happened (e.g. what warnings or errors you are getting)?

`no-extra-semicolons`
`block-closing-brace-newline-after` triggered

**Comment by jeddy3** [comment]
The `block-closing-brace-newline-after` stuff is a dupe of https://github.com/stylelint/stylelint/issues/1806. Lets makes this issue about `no-extra-semicolons`.

@stylelint/core We should probably update `no-extra-semicolons` so that it doesn't flag a single semicolon following a closing brace *of a custom property set block*. @evilebottnawi As you're closest to this rule, do you think this is can be done?

**Comment by evilebottnawi** [comment]
@jeddy3 let's create `isCustomSetsOfProperties` util(may be best name) and not trigger if it is?

**Comment by jeddy3** [comment]
Sgtm to me.

`isCustomPropertySet(rule)`

**Comment by kevinSuttle** [comment]
:thumbsup:

**Comment by evilebottnawi** [comment]
@jeddy3 @kevinSuttle not resolve, `postcss` parser invalid:
```css
:root { --foo: { margin: 0 }; --bar: { padding: 0 }; }
```
to
```shell
Rule {
  raws: { before: '', between: ' ', semicolon: false, after: '; ' },
  type: 'rule',
  nodes: 
   [ Rule {
       raws: [Object],
       type: 'rule',
       nodes: [Object],
       parent: [Circular],
       source: [Object],
       selector: '--foo:',
       lastEach: 1,
       indexes: {} },
     Rule {
       raws: [Object],
       type: 'rule',
       nodes: [Object],
       parent: [Circular],
       source: [Object],
       selector: '--bar:',
       lastEach: 1,
       indexes: {} } ],
  parent: 
   Root {
     raws: { semicolon: false, after: '' },
     type: 'root',
     nodes: [ [Circular] ],
     source: { input: [Object], start: [Object] },
     lastEach: 2,
     indexes: { '2': 0 } },
  source: 
   { start: { line: 1, column: 1 },
     input: 
      Input {
        css: ':root { --foo: { margin: 0; }; --bar: { padding: 0; }; }',
        id: '<input css 5>' },
     end: { line: 1, column: 56 } },
  selector: ':root',
  lastEach: 1,
  indexes: {} }
```
`--foo: { margin: 0; }`
to
```shell
Rule {
  raws: { before: ' ', between: ' ', semicolon: true, after: ' ' },
  type: 'rule',
  nodes: 
   [ Declaration {
       raws: [Object],
       type: 'decl',
       parent: [Circular],
       source: [Object],
       prop: 'margin',
       value: '0' } ],
  parent: 
   Rule {
     raws: { before: '', between: ' ', semicolon: false, after: '; ' },
     type: 'rule',
     nodes: [ [Circular], [Object] ],
     parent: 
      Root {
        raws: [Object],
        type: 'root',
        nodes: [Object],
        source: [Object],
        lastEach: 2,
        indexes: [Object] },
     source: { start: [Object], input: [Object], end: [Object] },
     selector: ':root',
     lastEach: 2,
     indexes: { '2': 0 } },
  source: 
   { start: { line: 1, column: 9 },
     input: 
      Input {
        css: ':root { --foo: { margin: 0; }; --bar: { padding: 0; }; }',
        id: '<input css 5>' },
     end: { line: 1, column: 29 } },
  selector: '--foo:',
  lastEach: 1,
  indexes: {} }
```
`--bar: { padding: 0; }`
to
```shell
Rule {
  raws: { before: '; ', between: ' ', semicolon: false, after: ' ' },
  type: 'rule',
  nodes: 
   [ Declaration {
       raws: [Object],
       type: 'decl',
       parent: [Circular],
       source: [Object],
       prop: 'padding',
       value: '0' } ],
  parent: 
   Rule {
     raws: { before: '', between: ' ', semicolon: false, after: '; ' },
     type: 'rule',
     nodes: [ [Object], [Circular] ],
     parent: 
      Root {
        raws: [Object],
        type: 'root',
        nodes: [Object],
        source: [Object],
        lastEach: 2,
        indexes: [Object] },
     source: { start: [Object], input: [Object], end: [Object] },
     selector: ':root',
     lastEach: 2,
     indexes: { '2': 1 } },
  source: 
   { start: { line: 1, column: 31 },
     input: 
      Input {
        css: ':root { --foo: { margin: 0 }; --bar: { padding: 0 }; }',
        id: '<input css 5>' },
     end: { line: 1, column: 51 } },
  selector: '--bar:',
  lastEach: 1,
  indexes: {} }
```
Problems in `raws`, `root` have `after` - `; `. But first `custom set` not have semicolon in `raws`, second have semicolon in `raws.before`.

If we use 
```css
:root { --foo: { margin: 0 }; --bar: { padding: 0 }; }
```
Second `custom set` will be have semicolon in `raws.after`.
I don't know is this normal behavior, let's wait what @ai would say on the subject.

**Comment by jeddy3** [comment]
@evilebottnawi Have you posted your parser findings to the postcss issue tracker?

**Comment by evilebottnawi** [comment]
@jeddy3 no, i mentioned @ai, but not received reply :smile: , tomorrow i will create issue and begin to deal with this problem, I'm interested in supporting `@apply` and all features in `stylelint`, especially new css features.

**Comment by ai** [comment]
Hi. I changed raws parser in 5.2. Could you test it again?

**Comment by evilebottnawi** [comment]
@ai thanks, today i will try

**Comment by web-padawan** [comment]
Any updates?

**Comment by evilebottnawi** [comment]
@web-padawan in the coming days I will try to solve this problem :smile: little free time

**Comment by web-padawan** [comment]
@evilebottnawi could you please take a look at #1806 as well? That issue is also related to CSS mixins.

**Comment by evilebottnawi** [comment]
@web-padawan alread in PR https://github.com/stylelint/stylelint/pull/1885, need review second eyes and then we will can merge, wait :smile:

**Comment by evilebottnawi** [comment]
@ai 
```css
Have css:
```css
:root { 
  --foo: { color: red }; 
  --bar: { color: blue }; 
  --foo-bar: { color: blue };
  --bar-foo: { color: blue }; 
}
```
Have js:
```js
root.walk((node) => {
      if (node.selector && node.selector.indexOf("--") !== -1) {
        console.log(node.raws)
      }
}
```
Output:
```shell
{ before: ' ', between: ' ', semicolon: false, after: ' ' }
{ before: '; ', between: ' ', semicolon: false, after: ' ' }
{ before: '; ', between: ' ', semicolon: false, after: ' ' }
{ before: '; ', between: ' ', semicolon: false, after: ' ' }
```
It is normal have `;` in `before`, not `after`?
Also first `custom-property-set` not have `;` in `before`, this semicolon in `after` into `root` node (`raws: { before: '', between: ' ', semicolon: false, after: '; ' }`).

**Comment by ai** [comment]
@evilebottnawi `Node#raws.after` is not about symbols after this node. It is a symbols before `}`:

```
<before> selector <between> { children <after> }
```

**Comment by evilebottnawi** [comment]
@ai thanks, then this is the correct behavior as I understand it?

**Comment by ai** [comment]
@evilebottnawi yeap `; ` should be in next rule `before` :)

**Comment by evilebottnawi** [comment]
Done in https://github.com/stylelint/stylelint/pull/1900
